README
Bean-Counter

This is a collection of scripts useful for the anal-retentive teacher grading students in a very particular way.

hwtotal totals the scores in the homework directory and stores them in cumulative.yaml. However you may want to convert the scores to percentages of an ideal score. hwtotal doesn't do this.

grades totals subscores for classwork, homework and exams and prints a report.

checkbeans identifies the sources of the classwork score that a student gets by virtue of being a member of a group and getting a group score.

beans adds points that a group has accumulated and are being processed, for addition to the group record.

classwork totals scores that students have gained either individually or as members of groups over the weeks of the course. The average score per week is 80/number of weeks, so the total average is 80.

cumulative. How is this different than hwtotal?

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the README
file from a module distribution so that people browsing the archive
can use it get an idea of the modules uses. It is usually a good idea
to provide version information here so that people can decide whether
fixes for the module are worth downloading.

INSTALLATION

To install this module, run the following commands:

    perl Build.PL
    ./Build
    ./Build test
    ./Build install


SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the perldoc command.

    perldoc Bean::Counter

You can also look for information at:

    Search CPAN
        http://search.cpan.org/dist/Bean-Counter

    CPAN Request Tracker:
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Bean-Counter

    AnnoCPAN, annotated CPAN documentation:
        http://annocpan.org/dist/Bean-Counter

    CPAN Ratings:
        http://cpanratings.perl.org/d/Bean-Counter

COPYRIGHT AND LICENCE

Copyright (C) 2006 Dr Bean

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.


Changes
---
version: 0.12
date: 三  6月 30 11:53:56 CST 2010
changes:
  - BUILDARGS to avoid users building Classwork, Approach objects
---
version: 0.11
date: 日  6月 20 14:35:38 CST 2010
changes:
  - Catalyst's Makefile.PL confusing cpantesters
  - Get PkgVersion to get versions right
---
version: 0.10
date: 六  6月 19 18:03:17 CST 2010
changes:
  - Grades, Classwork delegate approach to CompComp or Groupwork objects.
  - Try::Tiny exception blocks
  - Work on making grading CompComp, Exams easier with scripts.
0.07 2010  2月 01, 10時30分50秒
  - Apply Groupwork, CompComp or Jigsaw roles to Classwork role
0.07 日  6月 14 11:06:11 CST 2009
  - Extensive use of MooseX::Types type checking.
0.06 日  6月 14 11:06:11 CST 2009
  - Left Beans names in t/00-load.t, etc. Left out dependencies in classes.
0.05 日  6月 14 11:06:11 CST 2009
  - MAKAROW has Script package/module.
0.04 五  6月 12 21:33:48 CST 2009
  - Name change to Grades. Rewriting with Moosex::Declare
0.03 六  1月 24 10:33:37 CST 2009
  - Combine for web interface to scripts for students to see grades.
  - Use of Moose
0.02  2009-01-22 21:40:48
        - initial revision, generated by Catalyst
0.01   2004 - 2008
        First version, was scripts


lib/beans/Homework.pm
#!/usr/bin/perl

package Script;
use Moose;
with 'MooseX::Getopt';

has 'man' => (is => 'ro', isa => 'Bool');
has 'help' => (is => 'ro', isa => 'Bool');
has 'league' => (traits => ['Getopt'], is => 'ro', isa => 'Str',
		cmd_aliases => 'l',);
has 'player' => (traits => ['Getopt'], is => 'ro', isa => 'Str',
		cmd_aliases => 'p',);

package League;
use Moose;
use YAML qw/LoadFile DumpFile/;
use List::MoreUtils qw/any/;

extends 'Script';

has 'yaml' => (is => 'ro', isa => 'HashRef', lazy_build => 1);
sub _build_yaml {
		my ($instance) = @_;
		my $league = $instance->league;
		LoadFile "$league/league.yaml"
}

has 'name' => (is => 'ro', isa => 'Str', lazy_build => 1);
sub _build_name {
	my $self = shift;
	my $data = $self->yaml;
	$data->{league};
}

has 'hwdir' => (is => 'ro', isa => 'Str', lazy_build => 1);
sub _build_hwdir {
	my $self = shift;
	my $league = $self->league;
	my $data = $self->yaml;
	my $hwdir = $data->{hw} || "$league/homework"
}
has 'rounds' => (is => 'ro', isa => 'ArrayRef', lazy_build => 1);
sub _build_rounds {
	my $self = shift;
	my $hwdir = $self->hwdir;
	my @hw = glob "$hwdir/*.yaml";
	[ sort {$a<=>$b} map m/^$hwdir\/(\d+)\.yaml$/, @hw ];
}
has 'hw' => (is => 'ro', isa => 'HashRef', lazy_build => 1);
sub _build_hw {
	my $self = shift;
	my $hwdir = $self->hwdir;
	my $rounds = $self->rounds;
	+{ map { $_ => LoadFile "$hwdir/$_.yaml" } @$rounds };
}
has 'hwMax' => (is => 'ro', isa => 'Int', lazy => 1, default =>
				sub { shift->yaml->{hwMax} } );
has 'totalScores' => (is => 'ro', isa => 'HashRef', lazy_build => 1);
sub _build_totalScores {
	my $self = shift;
	my $hwdir = $self->hwdir;
	LoadFile "$hwdir/total.yaml";
}
has 'totalPercent' => (is => 'ro', isa => 'HashRef', lazy_build => 1);
sub _build_totalPercent {
	my $self = shift;
	my $hwdir = $self->hwdir;
	LoadFile "$hwdir/percent.yaml";
}
has 'members' => (is => 'ro', isa => 'ArrayRef', lazy_build => 1);
sub _build_members {
	my $self = shift;
	my $data = $self->yaml;
	my @members = sort { $a->{id} cmp $b->{id} } @{$data->{member}};
	[ map { Player->new(league => $self, id=>$_->{id},name=>$_->{name})
			} @members ];
	# $data->{member};
}

sub is_member {
	my $self = shift;
	my $id = shift;
	my $data = $self->yaml;
	any { $_->{id} eq $id } @{$data->{member}};
}

sub save {
	my $self = shift;
	DumpFile shift(), shift();
}
	


package Player;
use Moose;
extends 'League';
use List::MoreUtils qw/firstval/;

has 'league' => (is => 'ro', isa => 'League', required => 1);
has 'id' => (is => 'ro', isa => 'Str', required => 1);
has 'name' => (is => 'ro', isa => 'Str', lazy_build => 1);
sub _build_name {
	my $self = shift;
	my $league = $self->league;
	my $id = $self->id;
	my $members = $league->members;
	my $member = firstval { $_->id eq $id } @$members;
	$member->name;
}
has 'Chinese' => (is => 'ro', isa => 'Str');
has 'grades' => (is => 'ro', isa => 'ArrayRef', lazy_build => 1);
sub _build_grades {
	my $self = shift;
	my $id = $self->id;
	my $hw = $self->hw;
	my $rounds = $self->rounds;
	[ map { $hw->{$_}->{$id} } @$rounds ];
}

package main;

use strict;
use warnings;

use List::Util qw/sum/;
use IO::All;
use IO::Handle;
use Cwd;
use Pod::Usage;

run() unless caller;

sub run {
	my $script = Script->new_with_options( league => getcwd );
	pod2usage(1) if $script->help;
	pod2usage(-exitstatus => 0, -verbose => 2) if $script->man;
	my $leagueId = $script->league;
	my $league = League->new( league => $leagueId ) or
		die "No $leagueId league: $!";
	my $hwMax = $league->hwMax;
	my $rounds = $league->rounds;
	my $totalMax = @$rounds * $hwMax;
	my $oldtotal = $league->totalScores;
	my $oldpercent = $league->totalPercent;
	my ($newtotal, $newpercent);
	my @formats = qw/'' WEEKS1 WEEKS2 WEEKS3 WEEKS4 WEEKS5 WEEKS6 WEEKS7
		WEEKS8 WEEKS9 WEEKS10 WEEKS11 WEEKS12 WEEKS13/;
	REP->format_name($formats[@$rounds]);
	open REP, '>-' or die 'STDOUT? $!'; 
	my $name = $league->name;
	my @romans = qw/'' 01 02 03 04 05 06 07
		08 09 10 11 12 13 14 15 16 17 18/;
	local $,=', ';
	print REP
"             $name Homework
		    Weeks: Grade(0-$hwMax)/$hwMax
Name        ID  @romans[@$rounds]   Total/$totalMax Grade/100
-------------------------------------------------------------------------
";
	my $members = $league->is_member( $script->player )? 
		[ Player->new( id => $script->player, league => $league ) ]:
		$league->members;
	foreach my $member ( @$members ) 
	{
		our $id = $member->id;
		our $name = $member->name;
		# our $indgrades = [ (2) x @$rounds ];
		our @indgrades = @{ $member->grades };
		@indgrades = map { $_ == 0.5? substr $_, 1: $_ } @indgrades;
		die "No homework grade for $name $id" if grep {not defined} @indgrades;
		our $grade = sum @indgrades;
		our $percent = 100 * $grade / $totalMax;
		$newtotal->{$id} = $grade;
		$newpercent->{$id} = $percent;
		write REP;
		# $grades->{$partner} = $newGrade;
	} 
	
	
	my $hwdir = $league->hwdir;
	my $log = io "$hwdir/hwtotal.log";
	localtime() . ": $0 run to calculate homework totals at round $rounds->[-1].\n"
										>> $log;
	$league->save("$hwdir/total.yaml.bak", $oldtotal);
	$league->save("$hwdir/total.yaml", $newtotal);
	$league->save("$hwdir/percent.yaml.bak", $oldpercent);
	$league->save("$hwdir/percent.yaml", $newpercent);
}

our ($name, $id, @indgrades, $grade, $percent);
format WEEKS1 =
@<<<<<< @<<<<<<< @<<   @<<<<<   @<<
$name,  $id,	@indgrades, $grade, $percent
.
format WEEKS2 =
@<<<<<< @<<<<<<< @<<@<<   @<<<<<   @<<
$name,  $id,	@indgrades, $grade, $percent
.
format WEEKS3 =
@<<<<<< @<<<<<<< @<<@<<@<<   @<<<<<   @<<
$name,  $id,	@indgrades, $grade, $percent
.
format WEEKS4 =
@<<<<<< @<<<<<<< @<<@<<@<<@<<   @<<<<<   @<<
$name,  $id,	@indgrades, $grade, $percent
.
format WEEKS5 =
@<<<<<< @<<<<<<< @<<@<<@<<@<<@<<   @<<<<<   @<<
$name,  $id,	@indgrades, $grade, $percent
.
format WEEKS6 =
@<<<<<< @<<<<<<< @<<@<<@<<@<<@<<@<<   @<<<<<   @<<
$name,  $id,	@indgrades, $grade, $percent
.
format WEEKS7 =
@<<<<<< @<<<<<<< @<<@<<@<<@<<@<<@<<@<<   @<<<<<   @<<
$name,  $id,	@indgrades, $grade, $percent
.
format WEEKS8 =
@<<<<<< @<<<<<<< @<<@<<@<<@<<@<<@<<@<<@<<   @<<<<<   @<<
$name,  $id,	@indgrades, $grade, $percent
.
format WEEKS9 =
@<<<<<< @<<<<<<< @<<@<<@<<@<<@<<@<<@<<@<<@<<   @<<<<<   @<<
$name,  $id,	@indgrades, $grade, $percent
.
format WEEKS10 =
@<<<<<< @<<<<<<< @<<@<<@<<@<<@<<@<<@<<@<<@<<@<<   @<<<<<   @<<
$name,  $id,	@indgrades, $grade, $percent
.
format WEEKS11 =
@<<<<<< @<<<<<<< @<<@<<@<<@<<@<<@<<@<<@<<@<<@<<@<<   @<<<<<   @<<
$name,  $id,	@indgrades, $grade, $percent
.
format WEEKS12 =
@<<<<<< @<<<<<<< @<<@<<@<<@<<@<<@<<@<<@<<@<<@<<@<<@<<   @<<<<<   @<<
$name,  $id,	@indgrades, $grade, $percent
.
format WEEKS13 =
@<<<<<< @<<<<<<< @<<@<<@<<@<<@<<@<<@<<@<<@<<@<<@<<@<<@<<   @<<<<<   @<<
$name,  $id,	@indgrades, $grade, $percent
.
format WEEKS14 =
@<<<<<< @<<<<<<< @<<@<<@<<@<<@<<@<<@<<@<<@<<@<<@<<@<<@<<@<<   @<<<<<   @<<
$name,  $id,	@indgrades, $grade, $percent
.

1;

__END__

=head1 NAME

cumulative - Add player results in individual rounds to get a cumulative total and show present standing

=head1 SYNOPSIS

hwtotal [options] 

Options:

--help            This help message

--man            A man page

--league m/j	The league whose results these are

=head1 OPTIONS

=over 8

=item B<-league>

The league to which the redeemer belongs

=back

=head1 DESCRIPTION

B<hwtotal> tallies individuals' scores in the files in the hw directory recorded in league.yaml. It stores the total in cumulative.yaml, in the same directory.

=cut

